### PR TITLE
#142 Externally defined schedule implemented

### PR DESCRIPTION
**Pull request by YuraRiabov**
Screenshot_638
Screenshot_639
Screenshot_640
Screenshot_641
Screenshot_642

### CODE REVIEW

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 1]
```diff
@@ -217,8 +230,11 @@ public async Task<AvailabilitySlotDto> GetAvailabilitySlotById(long id)
 
                 _context.Update(availabilitySlot);
             }
-            
-            _mapper.Map(updateAvailabilityDto.Schedule, availabilitySlot.Schedule);
+
+            if (!updateAvailabilityDto.Schedule.WithTeamMembers)
+            {
+                _mapper.Map(updateAvailabilityDto.Schedule, availabilitySlot.SlotMembers.First().Schedule);
```
Comment (Erbidium) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:236]: Why we simply take schedule of the first slotmember?

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 2]
```diff
@@ -245,6 +261,39 @@ public async Task DeleteAvailabilitySlot(long slotId)
             await _context.SaveChangesAsync();
         }
 
+        public async Task<AvailabilitySlotDto?> GetByLink(string link)
+        {
+            var slot = await GetByLinkInternal(link);
+            return slot is null ? null : _mapper.Map<AvailabilitySlotDto>(slot);
```
Comment (Erbidium) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: You can use ?? operator here

File: frontend/src/app/core/helpers/app-domain-helper.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export const getAppDomain = (): string => {
```
Comment (Erbidium) [inline frontend/src/app/core/helpers/app-domain-helper.ts]: Probably you can use document.location.hostname to get appp domain?

File: frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<div class="container">
+    <div class="schedule-container">
+        <div class="schedule-select-container">
+            <p>Schedule</p>
+            <div class="input">
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html:5]: Probably this div is redundant

**Review by Erbidium — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 1 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:236]: If schedule isn't set to have teamMembers(which is hardcoded now) we only have one slot member(author)

File: frontend/src/app/core/helpers/app-domain-helper.ts (hunk 3 above)
Comment (YuraRiabov) [reply frontend/src/app/core/helpers/app-domain-helper.ts]: No, it will return localhost while we need http://localhost:4200 or localhost:4200

File: frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html (hunk 4 above)
Comment (YuraRiabov) [reply frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html:5]: I don't think so cause we have styles for it, this component is just a part of what we had in schedule component beforehand

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 2 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Fixed

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 5]
```diff
@@ -245,6 +261,39 @@ public async Task DeleteAvailabilitySlot(long slotId)
...
+                _mapper.Map(scheduleDto, member.Schedule);
+            }
+
+            await _context.SaveChangesAsync();
+        }
+
+        private async Task<AvailabilitySlot?> GetByLinkInternal(string link)
+        {
+            try
+            {
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: We have global error handler for this

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 6]
```diff
@@ -245,6 +261,39 @@ public async Task DeleteAvailabilitySlot(long slotId)
...
+                return await _context.AvailabilitySlots
+                    .Include(slot => slot.SlotMembers)
+                        .ThenInclude(slot => slot.Schedule)
+                            .ThenInclude(s => s.ScheduleItems)
+                    .SingleAsync(s => s.Link == link);
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: It would be better to use FirstOrDefault and check for null

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 7]
```diff
@@ -245,6 +261,39 @@ public async Task DeleteAvailabilitySlot(long slotId)
...
+        }
+
+        public async Task UpdateScheduleExternally(string link, ScheduleDto scheduleDto)
+        {
+            var slot = await GetByLinkInternal(link);
+            foreach (var member in slot!.SlotMembers)
+            {
+                _mapper.Map(scheduleDto, member.Schedule);
+            }
+
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:277]: You can map list in the auto mapper

File: frontend/src/app/core/helpers/app-domain-helper.ts (hunk 3 above)
Comment (tatianahutii) [reply frontend/src/app/core/helpers/app-domain-helper.ts]: why do we need this helper?

File: frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass [hunk 8]
```diff
@@ -19,8 +19,8 @@ app-side-menu-tabs
     display: flex
     flex-direction: column
     justify-items: center
-    padding: 0 100px 90px 100px
-    @media screen and (max-width: $desktop-min-width)
+    padding: 0 100px 0 100px
+    @media screen and(max-width: $desktop-min-width)
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass]: Why did you delete whitespace?

File: frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.ts [hunk 9]
```diff
@@ -45,6 +51,10 @@ export class NewAvailabilityComponent implements OnInit, AfterViewInit {
 
     index: number = 0;
 
+    link: string;
+
+    public appDomain = getAppDomain();
+
     hasAnyErrors: boolean = false;
 
     // eslint-disable-next-line no-empty-function
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.ts]: Please delete eslint-disable-next-line

File: frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                    [item]="schedule.scheduleItems[i]"
+                    [itemChange]="changeEvent"
+                    [displayDay]="displayDays[i]">
+                </app-schedule-list-item>
+            </mat-tab>
+            <mat-tab label="Week View" *ngIf="!disabled">
+                <app-schedule-week
+                    [items]="schedule.scheduleItems"
+                    [itemChange]="changeEvent"
+                >
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.html]:
```suggestion
                    [itemChange]="changeEvent">
```

File: frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.sass [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+.schedule-select-container
+    display: flex
+    justify-content: space-between
+    align-items: center
+    padding-bottom: 10px
+    p
+        margin: 0
+        font-weight: bold
+
+
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.sass]: Please delete

File: frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    readonly timeZones: ITimeZone[] = getPossibleTimeZones();
+
+    selectedTimeZone: string;
+
+    changeTimeZone() {
+        this.schedule.timeZone = this.getSelectedTimeZoneValue();
+    }
+
+    getSelectedTimeZoneValue() {
+        return this.timeZones.find((x) => x.displayValue === this.selectedTimeZone)!.value;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-definition/schedule-definition.component.ts]: Why do you always use ! ?

File: frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                    if (!slot) {
+                        this.notificationsService.showErrorMessage('Link is invalid or not unique');
+                        this.router.navigate(['/availability']);
+
+                        return;
+                    }
+                    this.slot = slot;
+                    this.schedule = slot.schedule;
+                    this.schedule.scheduleItems = getScheduleItems();
+                });
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts]: move this to a separate method

File: frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts [hunk 14]
```diff
@@ -1,48 +1,54 @@
...
+        this.schedule = this.slot?.schedule ?? getDefaultSchedule(false);
     }
 
-    changeEvent: EventEmitter<any> = new EventEmitter();
+    constructor(private notificationsService: NotificationService, private clipboard: Clipboard) {
+        super();
+    }
 
-    slot?: IAvailabilitySlot;
+    @Input() public slotLink: string;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts]: Move this before constructor

File: frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts [hunk 15]
```diff
@@ -1,48 +1,54 @@
...
+    public externalDefinitionToggle($event: MatSlideToggleChange) {
+        if (!this.schedule.definedExternally && !this.slotLink) {
+            this.notificationsService.showInfoMessage('Slot link must be defined and unique to allow external schedule definition');
+            $event.source.checked = false;
 
-    displayDays: string[] = getDisplayDays();
+            return;
+        }
+        this.schedule.definedExternally = !this.schedule.definedExternally;
+        this.schedule = getDefaultSchedule(this.schedule.definedExternally);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts]: Why we cannot combine this?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 16]
```diff
@@ -217,8 +230,11 @@ public async Task<AvailabilitySlotDto> GetAvailabilitySlotById(long id)
...
             }
-            
-            _mapper.Map(updateAvailabilityDto.Schedule, availabilitySlot.Schedule);
+
+            if (!updateAvailabilityDto.Schedule.WithTeamMembers)
+            {
+                _mapper.Map(updateAvailabilityDto.Schedule, availabilitySlot.SlotMembers.First().Schedule);
+            }
 
             availabilitySlot.LocationType = updateAvailabilityDto.GeneralDetails!.LocationType;
```
Comment (Erbidium) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:239]: Why ! is used here?

File: frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass [hunk 17]
```diff
@@ -117,3 +117,6 @@ app-side-menu-tabs
 .active-container
     display: flex
     align-items: center
+
+::ng-deep .menu-item-container .mat-tab-body-wrapper
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass:121]: Maybe we can avoid using ng deep))?

**Review by Erbidium — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 16 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:239]: If you mean line 234, cause we use SlotMembers.First() only when we know there is only one slot member aka author, if you mean 239, it's not my code, but it's used cause someone made general details nullable in the dto

File: frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass (hunk 17 above)
Comment (YuraRiabov) [reply frontend/src/app/modules/availability/new-slot/new-availability/new-availability.component.sass:121]: Unfortunately it's needed here, to have placeholder(appears when slot is set to be defined externally but isn't yet) cover the whole space available mat-tab-body-wrapper has to take all the space it has, which is achieved via flex-grow property

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 16 above)
Comment (Erbidium) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:239]: I meant line 239

**Review by Erbidium — APPROVED** [review]

File: frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+            .subscribe(slot => {
+                if (!slot) {
+                    this.notificationsService.showErrorMessage('Link is invalid or not unique');
+                    this.router.navigate(['/availability']);
+
+                    return;
+                }
+                this.slot = slot;
+                this.schedule = slot.schedule;
+                this.schedule.scheduleItems = getScheduleItems();
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts]:
```suggestion
                this.schedule = {
                    ...slot.schedule,
                    scheduleItems: getScheduleItems()
                }
```

File: frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        this.definingSchedule = false;
+        this.definingEmail = true;
+    }
+
+    public doneClicked() {
+        this.slotService.updateScheduleExternally(this.link, this.schedule)
+            .pipe(this.untilThis)
+            .subscribe(() => {
+                this.definingEmail = false;
+                this.done = true;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-external/schedule-external.component.ts]: Why do we need this variable?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Comment by tatianahutii** [comment]
Please squash your migrations

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]
